File: src/main/java/org/minefortress/entity/ai/goal/EatGoal.java
package org.minefortress.entity.ai.goal;

import net.minecraft.entity.ai.goal.Goal;
import net.remmintan.mods.minefortress.core.interfaces.entities.pawns.IFortressAwareEntity;
import net.remmintan.mods.minefortress.core.interfaces.entities.pawns.controls.IEatControl;
import net.remmintan.mods.minefortress.core.interfaces.resources.IServerResourceManager;
import net.remmintan.mods.minefortress.core.interfaces.server.IServerManagersProvider;
import net.remmintan.mods.minefortress.core.utils.ServerModUtils;
import org.jetbrains.annotations.NotNull;
import org.minefortress.entity.HungryEntity;

import java.util.Optional;

public class EatGoal extends Goal {
    private final HungryEntity entity;

    public EatGoal(HungryEntity entity) {
        this.entity = entity;
    }

    @Override
    public boolean canStart() {
        return getEatControl().isHungry() && getResourceManager().flatMap(IServerResourceManager::getEatableItem).isPresent();
    }

    @Override
    public void start() {
        getResourceManager().ifPresent(it ->
                it.getEatableItem().ifPresent(st -> {
                it.increaseItemAmount(st.getItem(), -1);
                getEatControl().eatFood(st.getItem());
            })
        );
    }

    @Override
    public boolean shouldContinue() {
        return super.shouldContinue() && getEatControl().isEating();
    }

    private Optional<IServerResourceManager> getResourceManager() {
        return entity instanceof IFortressAwareEntity fortressEntity ?
                ServerModUtils.getManagersProvider(fortressEntity)
                        .map(IServerManagersProvider::getResourceManager) :
                Optional.empty();
    }

    @NotNull
    private IEatControl getEatControl() {
        return entity.getEatControl().orElseThrow();
    }



}

File: src/main/java/org/minefortress/entity/ai/goal/EatGoal.kt
package org.minefortress.entity.ai.goal

import net.minecraft.entity.ai.goal.Goal
import net.remmintan.mods.minefortress.core.interfaces.entities.pawns.IFortressAwareEntity
import net.remmintan.mods.minefortress.core.interfaces.entities.pawns.controls.IEatControl
import net.remmintan.mods.minefortress.core.interfaces.resources.server.IServerFoodManager
import net.remmintan.mods.minefortress.core.utils.getManagersProvider
import org.minefortress.entity.HungryEntity

class EatGoal(private val entity: HungryEntity) : Goal() {
    override fun canStart(): Boolean {
        return eatControl.isHungry && foodManager.hasFood()
    }

    override fun start() {
        foodManager.getFood()?.let {
            eatControl.eatFood(it.item)
        }
    }

    override fun shouldContinue(): Boolean {
        return super.shouldContinue() && eatControl.isEating
    }

    private val foodManager: IServerFoodManager by lazy {
        val server = entity.server!!
        require(entity is IFortressAwareEntity)

        server.getManagersProvider(entity.fortressPos!!).foodManager
    }

    private val eatControl: IEatControl
        get() = entity.eatControl.orElseThrow()
}
